File: gui/src/pgn.ts
// PGN in and out. Writing is plain formatting. Reading only extracts the
// moves as written; which move each one is -- and whether it is legal at
// all -- is decided by the engine, from its `san` list for each position,
// so the GUI still implements no chess rules of its own.

export interface PgnInfo {
  sans: string[];
  white: string;
  black: string;
  result: string; // "1-0", "0-1", "1/2-1/2" or "*"
  eco?: string;
  opening?: string;
  timeControl?: string; // PGN form, e.g. "300+2"
}

function tag(name: string, value: string): string {
  return `[${name} "${value.replace(/\\/g, "\\\\").replace(/"/g, '\\"')}"]`;
}

export function toPgn(g: PgnInfo, date = new Date()): string {
  const d = `${date.getFullYear()}.${String(date.getMonth() + 1).padStart(2, "0")}.${String(date.getDate()).padStart(2, "0")}`;
  const tags = [
    tag("Event", "Simorgh game"),
    tag("Site", "Simorgh"),
    tag("Date", d),
    tag("Round", "-"),
    tag("White", g.white),
    tag("Black", g.black),
    tag("Result", g.result),
  ];
  if (g.eco) tags.push(tag("ECO", g.eco));
  if (g.opening) tags.push(tag("Opening", g.opening));
  if (g.timeControl) tags.push(tag("TimeControl", g.timeControl));

  // Movetext, wrapped at 80 columns as the PGN standard asks.
  const words: string[] = [];
  g.sans.forEach((san, i) => {
    if (i % 2 === 0) words.push(`${i / 2 + 1}.`);
    words.push(san);
  });
  words.push(g.result);
  const lines: string[] = [];
  let line = "";
  for (const w of words) {
    if (line && line.length + 1 + w.length > 80) {
      lines.push(line);
      line = w;
    } else line = line ? `${line} ${w}` : w;
  }
  if (line) lines.push(line);
  return `${tags.join("\n")}\n\n${lines.join("\n")}\n`;
}

/** The moves of the first game in `text`, as written (SAN), or an error. */
export function readPgnMoves(text: string): { sans: string[] } | { error: "setup" | "empty" } {
  // A game that starts from a set-up position cannot be replayed from the
  // initial position, which is the only start the apps know.
  if (/\[\s*SetUp\s+"1"\s*\]/i.test(text) || /\[\s*FEN\s+"/i.test(text)) return { error: "setup" };

  let body = text
    .split(/\r?\n/)
    .filter((l) => !/^\s*\[.*\]\s*$/.test(l)) // tag pairs
    .map((l) => l.replace(/;.*$/, "")) // rest-of-line comments
    .join(" ")
    .replace(/\{[^}]*\}/g, " ") // brace comments
    .replace(/\$\d+/g, " "); // numeric annotation glyphs

  // Variations nest, so strip them with a depth count, not a regex.
  let depth = 0;
  let flat = "";
  for (const ch of body) {
    if (ch === "(") depth++;
    else if (ch === ")") depth = Math.max(0, depth - 1);
    else if (depth === 0) flat += ch;
  }
  body = flat;

  const sans: string[] = [];
  for (let tok of body.split(/\s+/)) {
    tok = tok.replace(/^\d+\.+/, ""); // "12." "12..." and "12.e4"
    if (!tok) continue;
    if (/^(1-0|0-1|1\/2-1\/2|\*)$/.test(tok)) break; // end of the game
    sans.push(tok);
  }
  return sans.length ? { sans } : { error: "empty" };
}

/**
 * The same move whether written "Nf3+", "Nf3!?", "0-0", "e8Q" or "e8=Q":
 * annotations and check marks are dropped, and the common variants of
 * castling and promotion are brought to one form.
 */
export function normalizeSan(san: string): string {
  return san
    .replace(/[+#!?]+$/g, "")
    .replace(/e\.p\.$/, "")
    .replace(/^0-0-0/, "O-O-O")
    .replace(/^0-0/, "O-O")
    .replace(/([a-h][18])([QRBN])$/, "$1=$2");
}

File: android/app/src/main/java/ir/simorgh/chess/Pgn.kt
package ir.simorgh.chess

import java.util.Calendar

/**
 * PGN in and out, the same as the desktop app's pgn.ts. Writing is plain
 * formatting. Reading only extracts the moves as written; which move each
 * one is, and whether it is legal at all, the engine decides from its `san`
 * list for each position, so this app still implements no chess rules.
 *
 * Every brace and bracket in the patterns below is escaped, even where
 * desktop Java would let it pass: Android compiles regexes with ICU, which
 * rejects an unescaped `}` -- the first version of this file crashed on
 * import because of exactly that.
 */
object Pgn {

    class Info(
        val sans: List<String>,
        val white: String,
        val black: String,
        val result: String,          // "1-0", "0-1", "1/2-1/2" or "*"
        val eco: String? = null,
        val opening: String? = null,
        val timeControl: String? = null,
    )

    private fun tag(name: String, value: String) =
        "[$name \"${value.replace("\\", "\\\\").replace("\"", "\\\"")}\"]"

    fun write(g: Info): String {
        val c = Calendar.getInstance()
        val date = "%04d.%02d.%02d".format(c.get(Calendar.YEAR), c.get(Calendar.MONTH) + 1, c.get(Calendar.DAY_OF_MONTH))
        val tags = mutableListOf(
            tag("Event", "Simorgh game"), tag("Site", "Simorgh"), tag("Date", date), tag("Round", "-"),
            tag("White", g.white), tag("Black", g.black), tag("Result", g.result),
        )
        g.eco?.let { tags += tag("ECO", it) }
        g.opening?.let { tags += tag("Opening", it) }
        g.timeControl?.let { tags += tag("TimeControl", it) }

        // Movetext, wrapped at 80 columns as the PGN standard asks.
        val words = mutableListOf<String>()
        g.sans.forEachIndexed { i, san ->
            if (i % 2 == 0) words += "${i / 2 + 1}."
            words += san
        }
        words += g.result
        val lines = mutableListOf<String>()
        var line = ""
        for (w in words) {
            line = when {
                line.isEmpty() -> w
                line.length + 1 + w.length > 80 -> { lines += line; w }
                else -> "$line $w"
            }
        }
        if (line.isNotEmpty()) lines += line
        return tags.joinToString("\n") + "\n\n" + lines.joinToString("\n") + "\n"
    }

    sealed class Read {
        class Moves(val sans: List<String>) : Read()
        object SetUp : Read()   // starts from a set-up position
        object Empty : Read()
    }

    /** The moves of the first game in `text`, as written. */
    fun readMoves(text: String): Read {
        if (Regex("""\[\s*SetUp\s+"1"\s*\]""", RegexOption.IGNORE_CASE).containsMatchIn(text) ||
            Regex("""\[\s*FEN\s+"""", RegexOption.IGNORE_CASE).containsMatchIn(text)) return Read.SetUp

        var body = text.lines()
            .filterNot { Regex("""^\s*\[.*\]\s*$""").matches(it) }   // tag pairs
            .joinToString(" ") { it.replace(Regex(";.*$"), "") }   // rest-of-line comments
            .replace(Regex("""\{[^\}]*\}"""), " ")                    // brace comments
            .replace(Regex("""\$\d+"""), " ")                       // annotation glyphs

        // Variations nest, so strip them with a depth count, not a regex.
        val flat = StringBuilder()
        var depth = 0
        for (ch in body) when {
            ch == '(' -> depth++
            ch == ')' -> depth = maxOf(0, depth - 1)
            depth == 0 -> flat.append(ch)
        }
        body = flat.toString()

        val sans = mutableListOf<String>()
        for (raw in body.split(Regex("\\s+"))) {
            val tok = raw.replace(Regex("""^\d+\.+"""), "")
            if (tok.isEmpty()) continue
            if (Regex("""^(1-0|0-1|1/2-1/2|\*)$""").matches(tok)) break
            sans += tok
        }
        return if (sans.isEmpty()) Read.Empty else Read.Moves(sans)
    }

    /** One form for "Nf3+", "Nf3!?", "0-0", "e8Q" and "e8=Q" alike. */
    fun normalize(san: String): String = san
        .replace(Regex("[+#!?]+$"), "")
        .replace(Regex("e\\.p\\.$"), "")
        .replace(Regex("^0-0-0"), "O-O-O")
        .replace(Regex("^0-0"), "O-O")
        .replace(Regex("([a-h][18])([QRBN])$"), "$1=$2")
}
